install/first_halyard_boot.sh

```diff
@@ -0,0 +1,202 @@
+#!/bin/bash
+#
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script is specific to preparing a Google-hosted virtual machine
+# for running Spinnaker when the instance was created with metadata
+# holding configuration information, using Halyard.
+
+set -e
+set -u
+
+METADATA_URL="http://metadata.google.internal/computeMetadata/v1"
+INSTANCE_METADATA_URL="$METADATA_URL/instance"
+KUBE_DIR="/home/spinnaker/.kube"
+GCR_DIR="/home/spinnaker/.gcr"
+GCE_DIR="/home/spinnaker/.gce"
+
+DOCKER_ACCOUNT=my-gcr-account
+KUBERNETES_ACCOUNT=my-k8s-account
+GOOGLE_ACCOUNT=my-google-account
+
+function get_instance_metadata_attribute() {
+  local name="$1"
+  local value=$(curl -s -f -H "Metadata-Flavor: Google" \
+                     $INSTANCE_METADATA_URL/attributes/$name)
+  if [[ $? -eq 0 ]]; then
+    echo "$value"
+  else
+    echo ""
+  fi
+}
+
+function clear_metadata_to_file() {
+  local key="$1"
+  local path="$2"
+  local value=$(get_instance_metadata_attribute "$key")
+
+  if [[ $value = *[![:space:]]* ]]; then
+     echo "$value" > $path
+     chown spinnaker:spinnaker $path
+     clear_instance_metadata "$key"
+     if [[ $? -ne 0 ]]; then
+       die "Could not clear metadata from $key"
+     fi
+     return 0
+  elif [[ $value != "" ]]; then
+     # Clear key anyway, but act as if it werent there.
+     clear_instance_metadata "$key"
+  fi
+
+  return 1
+}
+
+function clear_instance_metadata() {
+  gcloud compute instances remove-metadata `hostname` \
+      --zone $MY_ZONE \
+      --keys "$1"
+  return $?
+}
+
+function replace_startup_script() {
+  # Keep the original around for reference.
+  # From now on, all we need to do is start_spinnaker
+  local original=$(get_instance_metadata_attribute "startup-script")
+  echo "$original" > "$SPINNAKER_INSTALL_DIR/scripts/original_startup_script.sh"
+  clear_instance_metadata "startup-script"
+}
+
+function install_halyard() {
+  wget https://raw.githubusercontent.com/spinnaker/halyard/master/InstallHalyard.sh
```

Comment:
This means to test a new halyard you need to either build a different version of the bootscript that changes this or you need to create the VM without this as the startscript, then login, edit the script, and then manually run it.

Both these can still be automated but perhaps this could be parameterized.
The simplest thing to do would be to let someone pass the script through metadata then have this first check the metadata then do the wget. Now it's the same typical API but you have a hook to facilitate development and testing.

Also, why wget in some places and curl in others?